### PR TITLE
AP_DDS: Mode Switch Service

### PR DESCRIPTION
**Pull request by arshPratap**
This PR aims to add Mode Switch service and has a dependency on #24705 
@srmainwaring @Ryanf55 @tridge

### CODE REVIEW

**Comment by Ryanf55** [comment]
Nice! Where is the mode switch IDL at? I would like to review the message and compare to rep147

**Comment by arshPratap** [comment]
@Ryanf55 you can find the IDLs currently defined in the yaml file

**Comment by Ryanf55** [comment]
> @Ryanf55 you can find the IDLs currently defined in the yaml file

Is there a way to add enums so the user knows what the modes ID's are?

**Comment by srmainwaring** [comment]
> Where is the mode switch IDL at?

The service interface definition is in `./Tools/ros2/ardupilot_msgs`. Perhaps as a follow up PR we could generate the IDL from this same as for other ROS msgs and then update the size and serialisation code?

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 1]
```diff
@@ -538,6 +539,42 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
...
+
+        ucdr_init_buffer(&reply_ub, reply_buffer, sizeof(reply_buffer));
+        bool serialize_success = true;
+        serialize_success &= ucdr_serialize_bool(&reply_ub, status);
+        serialize_success &= ucdr_serialize_uint8_t(&reply_ub, curr_mode);
+        if (serialize_success == false || reply_ub.error) {
+            break;
+        }
+
+        request_sample_count++;
```
Comment (tridge) [inline libraries/AP_DDS/AP_DDS_Client.cpp:568]: what is this for?

**Comment by arshPratap** [comment]
@srmainwaring done !

> 1. Update `./libraries/AP_DDS/README.md`
> 
> https://github.com/arshPratap/ardupilot/blob/4ec66c885446ab5fab7ae294f37aa43654d73826/libraries/AP_DDS/README.md?plain=1#L260
> 
> The name of the IS config file has changed to `DDS_Service_IS_config.yaml`.
> 
> 2. Provide an example for the mode switch service.

**Comment by srmainwaring** [comment]
Have replicated the examples. FWIW on macOS there are a few extra / more detailed steps needed to build and run.

### macOS instructions

Build `is-ros2-mix-generator`: 

```bash
colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_MACOSX_RPATH=FALSE -DUAGENT_USE_SYSTEM_FASTDDS=ON -DUAGENT_USE_SYSTEM_FASTCDR=ON -DMIX_ROS_PACKAGES="example_interfaces ardupilot_msgs" --packages-select is-ros2-mix-generator
```

Run `integration-service`:

```bash
LD_LIBRARY_PATH=$(pwd)/install/is-fastdds/lib ./build/is-core/integration-service ./src/ardupilot/libraries/AP_DDS/Is-Config/DDS_Service_IS_config.yaml --is-prefix-path=$(pwd)/install/is-ros2/lib/is/ros2 --is-prefix-path=$(pwd)/install/is-ros2-mix-generator/lib/is
```

Where I've placed the integration service package into the colcon workspace for ros2-ardupilot which reduces the need to switch between multiple projects and workspaces.

File: libraries/AP_DDS/AP_DDS_Client.cpp [hunk 2]
```diff
@@ -10,6 +10,7 @@
 #include <AP_BattMonitor/AP_BattMonitor.h>
 #include <AP_AHRS/AP_AHRS.h>
 #include <AP_Arming/AP_Arming.h>
+#include <AP_Vehicle/AP_Vehicle.h>
```
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:13]: The one thing I am unsure about is the circular includes. 

Oddly enough, AP_Vehicle.h doesn't include AP_DDS_Client.h.

Comment (srmainwaring) [reply libraries/AP_DDS/AP_DDS_Client.cpp:13]: `AP_Vehicle.h` forward declares `AP_DDS_Client`, has a pointer member variable `dds_client`, and pulls the include into the translation unit. So I think that's ok.

**Comment by arshPratap** [comment]
@srmainwaring thanks for the work on #24880 .. have rebased this PR over the latest changes..Hoping to have this merged soon.

File: libraries/AP_DDS/AP_DDS_Client.cpp (hunk 1 above)
Comment (srmainwaring) [reply libraries/AP_DDS/AP_DDS_Client.cpp]: This is from the example. I think it is intended for use with the `uxr_run_session_until_all_status` version of the `uxr_run_session_*` functions, where you'd capture the `request_id` for each topic / replier when serialising into the buffer.

**Comment by srmainwaring** [comment]
> have rebased this PR over the latest changes.

LGTM @arshPratap. Tested in SITL and working fine.

**Review by srmainwaring — APPROVED** [review]

**Review by Ryanf55 — APPROVED** [review]

**Pull request merged**